### PR TITLE
[rfc] snowflake observable source asset factory

### PR DESCRIPTION
**Pull request by dpeng817**
Adds an observable source asset for snowflake table freshness. Similarly to bigquery PR, goal here is mostly to evaluate the approach taken for retrieving last updated time, and refine into a resource method in a later PR. Starting to think we can introduce a configurable resource subclass for "freshness enabled" stuff, could potentially be a cool avenue.

How I tested this
This takes advantage of existing secrets in the buildkite bucket to test against live snowflake data. To my knowledge we don't do this in other tests. It's a pattern I stole from dagster-bigquery that I found nice, and think we should adopt here.

### CODE REVIEW

**Comment by dpeng817** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19700?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #19695** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19695?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #19698** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19698?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #19700** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19700?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
      * **PR #19691** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19691?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/19700?utm_source=stack-comment).

**Review by sryza — COMMENTED** [review]
I think we should start with a utility on the `SnowflakeResource` that fetches the last update timestamp for a set of tables.

I suspect that, in many cases, users are going to want to fetch other metadata alongside last update timestamp. But we can't make factory methods with every combination.  Also, the factory would need to pass through most of the attributes on `@observable_source_asset` to be broadly useful.

This makes me think that it's better to provide the factory as an example that people can copy and modify to their tastes, and instead focus on providing utilities that make the code required to build the factory as brief as possible.

**Comment by dpeng817** [comment]
Totally agreed, don't expect this to land, but I think it's useful to just see a working code snippet of an observable source asset querying live data, and use that as a jumping off point for building landable PRs (I allude to this in the description)

File: python_modules/libraries/dagster-snowflake/dagster_snowflake/assets.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            """
+            with conn.cursor() as cursor:
+                result = cursor.execute(query)
+                if not result:
+                    raise ValueError("No results found")
+                results = result.fetchall()
+
+        # context.log.info the last modified time as a UTC timestamp.
+        context.log.info(f"Last modified time for table {table}: {results[0][1]}")
+        return DataVersion(results[0][1].strftime("%Y-%m-%d %H:%M:%S.%f"))
```
Comment (sryza) [inline python_modules/libraries/dagster-snowflake/dagster_snowflake/assets.py:30]: I wouldn't necessarily characterize this as a data version. I think we should include it on some other property on the `ObserveResult`, probably metadata.

**Pull request closed without merging**